pydantic-ai-graph - simplify public generics

dmontagu opened the pull request:
This is a refactoring of the work in #528 attempting to reduce the number of user-facing generic parameters.

There are some things that we can probably adjust/remove, such as the introduction of _both_ `GraphRunner` and `GraphRun`. But there are some benefits to this, as it provides a way to get nearly the same API we currently have without BaseNode being aware of the input type, and without forcing BaseNode subclasses to have a fixed signature.

Some of the less consequential changes in here (such as moving/duplicating typevars) were done because I was hitting cyclic import errors and/or spurious type errors (I think I maybe ran into some bugs in pyright that I no longer know how to reproduce).

I'm happy to rework the implementation based on feedback if there's interest, but I don't really want to go through adding lots of tests and documenting everything if this line of implementation is going to be rejected, so @samuelcolvin maybe we can discuss synchronously.

dmontagu commented:
Note I say "user-facing generic parameters" because, for example, while `GraphRunner` has a `ParamSpec` as one of its parameters, users should never need to manually annotate it — they would just pass a node class to the `get_runner` method of `Graph`. This is important because it's hard to manually specify a paramspec for a node, but it's not so hard to manually specify the parameters for the `Graph` type itself, which may be desirable to do as a way to ensure the nodes being added satisfy constraints on their run-end types, rather than letting that be inferred by the type system.

dmontagu commented on pydantic_ai_graph/pydantic_ai_graph/state.py:
```diff
@@ -19,33 +24,40 @@ def serialize(self) -> bytes | None:
 
 
 @dataclass
-class Snapshot:
-    """Snapshot of a graph."""
-
-    last_node_id: str
-    next_node_id: str
-    start_ts: datetime
-    duration: float
-    state: bytes | None = None
```
I replaced this with non-serialized history events mostly because I think people will want to interact with the deserialized history, so storing it as raw bytes seems unfortunate. I think we need to ensure both `Step` and `EndEvent` are serializable (and therefore all `BaseNode` instances and all `RunEndT` values need to be serializable). I'm not sure the best way to achieve that though.

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/nodes.py:
```diff
@@ -1,113 +1,93 @@
 
-NodeInputT = TypeVar('NodeInputT', default=Any)
-GraphOutputT = TypeVar('GraphOutputT', default=Any)
+RunEndT = TypeVar('RunEndT', default=None)
```
I think this should not have a default.

dmontagu replied:
If we get rid of the default, then it needs to come before StateT everywhere. I'm okay with that, but just pointing it out. I will make that change

dmontagu replied:
Actually, it's awkward because, since parameters with defaults need to go after parameters without defaults, unless we drop the default value of `StateT` for `Graph`, we'd have to put `RunEndT` before `StateT` on `Graph`. But then you have to make a decision of whether you want consistency in parameter ordering on `BaseNode`, or convenience when omitting default parameters:

* If we keep the ordering of parameters on `BaseNode` as it currently is, then `Graph` has the parameters in one order and `BaseNode` has it in the other, which feels like a recipe for confusion.
* If we change the ordering of parameters on `BaseNode`, then to explicitly specify the `StateT` type you need to specify the `NodeRunEndT`, but explicitly specifying the `StateT` will be a lot more common than explicitly specifying the node run end type, so that feels pretty unfortunate.

Given the above, I now feel like we should _either_:
* keep a default of `None` for the `RunEndT` on `Graph`, or
* drop the default of `None` for the `StateT` on `Graph`.

I'm okay with either, lmk what you think. (I'm also okay with the first two bullets' suggestions, if you really prefer, but I think they are worse than making it so the graph typevars either both have defaults or both don't, and preserving parameter ordering).

dmontagu replied:
I'll note that trying to do the refactor, I think things get a little more complicated if we remove the default, and I'm not sure that having the default of `None` is that crazy given that it provides a way to end the run without returning a meaningful value, which may be more common than you'd expect if it ends up being common to create agents that do mutating tool calls before finishing rather than returning values in the end.

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/nodes.py:
```diff
@@ -1,113 +1,93 @@
-
-    def __init__(self, input_data: NodeInputT) -> None:
-        self.input_data = input_data
+    _node_id: ClassVar[str | None] = None
```
does this have to be private? I get it can be but think it'll be clearer that it's part of the public api if it's public.

dmontagu replied:
It can be part of the public API, but the problem is that in lots of places we were accessing it directly, rather than via `get_id()` which will grab the actual string, and which needs to be used everywhere (or else you'll get `None`s and therefore not unique ids). I changed it to private to make that mistake less likely but I understand your point that now it looks like users aren't supposed to set it. Maybe we can call it `node_id_override` instead? Or we can just keep it as `node_id` but basically I just don't want to make that mistake of accessing that value instead of calling `node.get_id()`. Maybe we can set it in the metaclass if it doesn't have an explicitly-set value?

dmontagu replied:
I wonder if it wouldn't just be better to drop the attribute entirely and have people override `get_id` if they want to manually change the value. That would ensure it's not possible to use the wrong api, for the in-my-opinion small cost of requiring a little more verbosity if you want to override the value. Given I expect we'll generally discourage overriding the value that seems okay to me

dmontagu replied:
I'm imagining adding a diff that looks like this:
```diff
     @classmethod
     @cache
     def get_id(cls) -> str:
-        return cls._node_id or cls.__name__
+        """Get the ID of the node.
+
+        You can override this if you want to serialize/deserialize nodes using a specific ID.
+        """
+        return cls.__name__
 ```

dmontagu replied:
I removed `_node_id` for now, didn't add that docstring since we don't have node serialization/deserialization now and not sure that's how it will work. Can undo this change if preferred though.

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/graph.py:
```diff
@@ -2,169 +2,232 @@
+        return next_node
+
+
+def mermaid_code(
```
we should move all this into a separate module (not now).

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/graph.py:
```diff
@@ -2,169 +2,232 @@
+            for _, next_node_id in sorted((node_order[node_id], node_id) for node_id in node_def.next_node_ids):
+                lines.append(f'  {node_id} --> {next_node_id}')
+        if node_def.returns_end:
+            lines.append(f'  {node_id} --> END')
```
this should match the node ID of `End`, I think `END` is better here, but I don't really mind.

dmontagu replied:
Just to make sure I understand correctly, are you suggesting we change:
`pydantic_ai_graph.nodes.End.get_id` to return `'END'` instead of `'End'`? I think that makes sense, but I'll also note that that method is unused and can probably just be deleted. At least, if I delete it, it doesn't cause any type checking errors.

dmontagu replied:
I removed `End.get_id()` for now

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/graph.py:
```diff
@@ -2,169 +2,232 @@
+    start: StartNodeProtocol[..., Any, Any] | tuple[StartNodeProtocol[..., Any, Any], ...] = (),
+    mermaid_ink_params: dict[str, str | int] | None = None,
+) -> None:
+    # TODO: do something with the path file extension, e.g. error if it's incompatible, or use it to specify a param
```
I think we remove `mermaid_ink_params` and just list out the params, then we can respect `type` from https://mermaid.ink/

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/graph.py:
```diff
@@ -2,169 +2,232 @@
-        *other_nodes: type[BaseNode[StateT, Any, GraphOutputT]],
-        name: str | None = None,
+        name: str = 'graph',
+        nodes: tuple[type[BaseNode[StateT, RunEndT]], ...] = (),
```
I would prefer `*nodes: BaseNode[StateT, RunEndT]]`

dmontagu replied:
There were a couple reasons I changed this:
* I liked having the name before the nodes (though I didn't realize the plan was to infer it from the assignment statement like with agent, which reduces that issue).
* I like being able to see where the top-level items start/end more clearly. Using an actual tuple here ensures they are indented relative to the other top-level kwargs in `Graph`. Without that, you get a bunch of items listed and then top-level kwargs mixed in.

	Concretely, I just find it visually easier to parse:
	```python
	g = Graph(
	    nodes=(
	        Float2String,
	        String2Length,
	        Double,
	    ),
	    state_type=MyState,
	)
	```
	vs.
	```python
	g = Graph(
	    Float2String,
	    String2Length,
	    Double,
	    state_type=MyState,
	)
	```
	Don't you? (And this only gets worse as you get more node types and more kwargs.)
	
	But I can revert this change if you insist.

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/graph.py:
```diff
@@ -2,169 +2,232 @@
-        first_node: type[BaseNode[StateT, GraphInputT, GraphOutputT]],
-        *other_nodes: type[BaseNode[StateT, Any, GraphOutputT]],
-        name: str | None = None,
+        name: str = 'graph',
```
we'll want the same logic as on agents where we infer the name from the frame, hence this should be be `None` by default.

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/graph.py:
```diff
@@ -2,169 +2,232 @@
-        for node in other_nodes:
-            node_def = node.get_node_def(parent_namespace)
-            nodes[node_def.node_id] = node_def
+        _node_defs: dict[str, NodeDef[StateT, RunEndT]] = {}
```
```suggestion
        self.node_defs: dict[str, NodeDef[StateT, RunEndT]] = {}
```

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/graph.py:
```diff
@@ -2,169 +2,232 @@
-        self.name = name
+        self._validate_edges()
+
+    def validate_next_node(self, next_node: BaseNode[StateT, RunEndT]) -> None:
```
this isn't used, is it necessary? If so, what's it for?

dmontagu replied:
I was using it, then broke it, but basically I wanted to set things up so that during a run it would check dynamically that the node was in the graph. But I guess I'm not sure if it's better to raise an error if you return something not in the expected graph or just chug along unimpacted given it doesn't _need_ to error. Maybe a warning is a good compromise?

dmontagu replied:
Specifically, I was using it in a callback, but then removed the callbacks stuff I had added since I think there's a decent amount of discussion to be had about the exact implementation details

dmontagu replied:
I removed this method for now

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/graph.py:
```diff
@@ -2,169 +2,232 @@
+        self, state: StateT, node: BaseNode[StateT, RunEndT]
+    ) -> tuple[RunEndT, list[GraphHistoryItem[StateT, RunEndT]]]:
+        if not isinstance(node, self.nodes):
+            raise ValueError(f'Node "{node.get_id()}" is not in the graph.')
```
```suggestion
            raise ValueError(f'Node "{node}" is not in the graph.')
```

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/graph.py:
```diff
@@ -2,169 +2,232 @@
+
+
+@dataclass
+class GraphRunner(Generic[RunSignatureT, StateT, RunEndT]):
```
I think this looks good, but I wonder if we could find a better name?

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/graph.py:
```diff
@@ -2,169 +2,232 @@
+
+
+@dataclass
+class GraphRun(Generic[StateT, RunEndT]):
```
are we sure we need this? Maybe we could make it private if we're not sure?

dmontagu replied:
It's currently the only thing that exposes the `step` API, which I feel like you want for unit testing. Note that I think we want to expose a way to do unit testing that goes beyond just calling `node.run` specifically for the sake of testing callbacks (once we add support for them).

But I think we could rework so `GraphRun` is private and there is a public function that users can call that basically does the same thing as `pydantic_ai_graph.graph.GraphRun.step` with callback execution, but doesn't require access to that class.

dmontagu replied:
To be clear I definitely don't think we need to have this class, it could be replaced with function calls operating on `state` and `history` directly, but:
* My comment above was a response to the suggestion of making it private — I think the functionality for stepping through a graph (with callbacks) should be public. (But that doesn't mean it needs to be a class.)
* I think having a class will make it easier to add features over time (in particular callbacks, though maybe other things). But we could drop the class for now and/or make it private as a way to keep users from depending on APIs we may want to break in the near future.

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/state.py:
```diff
@@ -19,33 +24,40 @@ def serialize(self) -> bytes | None:
+        self.state = copy.deepcopy(self.state)
+
+
+GraphHistoryItem = Union[Step[StateT, RunEndT], EndEvent[StateT, RunEndT]]
```
```suggestion
StepOrEnd = Union[Step[StateT, RunEndT], EndEvent[StateT, RunEndT]]
```
might be clearer.

dmontagu replied:
I just wasn't sure if we might add more over time, but I'm okay with that renaming if you are

dmontagu replied:
I'll rename it like this for now, if you foresee adding anything besides these though I think we should rename it now

samuelcolvin commented on tests/test_graph.py:
```diff
@@ -1,98 +1,115 @@
 async def test_graph():
-    class Float2String(BaseNode[None, float]):
+    @dataclass
+    class Float2String(BaseNode[None]):
```
```suggestion
    class Float2String(BaseNode):
```
Elegantly, we can remove the state generic param in these cases.

samuelcolvin reviewed (commented):
overall I think this looks good.

dmontagu commented:
I addressed some of the comments above, I've left unresolved the comments that I wasn't confident enough to resolve without further discussion (or at least further insistence on your part).
